### PR TITLE
Select device/host based atomics

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2393/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1730/ for details (Mellanox internal link).

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -19,6 +19,11 @@
 
 ucp_am_handler_t ucp_am_handlers[UCP_AM_ID_LAST] = {{0, NULL, NULL}};
 
+static const char *ucp_atomic_modes[] = {
+    [UCP_ATOMIC_MODE_HOST]   = "host",
```
Comment (shamisp) [inline src/ucp/core/ucp_context.c]: HOST == CPU ?

File: src/ucp/core/ucp_context.c [hunk 2]
```diff
@@ -84,12 +89,19 @@ static ucs_config_field_t ucp_config_table[] = {
    "Estimation of buffer copy bandwidth",
    ucs_offsetof(ucp_config_t, ctx.bcopy_bw), UCS_CONFIG_TYPE_MEMUNITS},
 
+  {"ATOMIC_MODE", "device",
+   "Atomic operations synchronization mode.\n"
+   " host   - operations are atomic with respect to the host processor.\n"
```
Comment (shamisp) [inline src/ucp/core/ucp_context.c]: I'm not sure if "host" is a good name ...

Comment (shamisp) [inline src/ucp/core/ucp_context.c]: Also, we have to specify what is the default

Comment (yosefe) [inline src/ucp/core/ucp_context.c]: the default is "device"

File: src/ucp/core/ucp_context.c (hunk 1 above)
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: yes

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2399/ for details.

File: src/ucp/core/ucp_context.c (hunk 2 above)
Comment (shamisp) [inline src/ucp/core/ucp_context.c]: pls. mention this explicitly in the help line

Comment (shamisp) [inline src/ucp/core/ucp_context.c]: name - what is used in verbs ? host or cpu ? they have similar parameter

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1735/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2404/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1740/ for details (Mellanox internal link).

File: src/ucp/core/ucp_context.c (hunk 2 above)
Comment (yosefe) [inline src/ucp/core/ucp_context.c]: @shamisp the default value would be printed in ucx_info explicitly, no need to add it again:
```
#
# Atomic operations synchronization mode.
#  host   - operations are atomic with respect to the host processor.
#  device - atomic operations are performed on one of the transport devices,
#           and there is no atomicity guarantee with respect to the host processor.
#
# Syntax: [host|device]
#
UCX_ATOMIC_MODE=device
```

Comment (yosefe) [inline src/ucp/core/ucp_context.c]: the terminology in verbs is "HCA/CPU"
however i prefer to have a more generic terminology in UCP

Comment (shamisp) [inline src/ucp/core/ucp_context.c]: when I think about host, I think about CPU, GPGPU, etc and then the term becomes confusing. It really depends on what memory you operate and where the memory controller resides...This is not a blocker, but I want to make the point

Comment (shamisp) [inline src/ucp/core/ucp_context.c]: BTW device also resided on the host ...

**Comment by shamisp** [comment]
@yosefe from technical perspective the code looks good. I'm only not sure about naming convention. Maybe this is something to discuss  (briefly) on the list, since we have a bit more users there. I want to make sure that we don't confuse people since this is only the parameter that we don't know to peak automatically

**Comment by hppritcha** [comment]
I think 'Host' might be a little ambiguous.  If in the past HCA/CPU was the convention, I'd vote for using 'CPU'.

**Comment by yosefe** [comment]
@shamisp renamed to 'cpu'

**Comment by shamisp** [comment]
👍

**Comment by shamisp** [comment]
@MattBBaker can you please confirm that it is functional on cray

**Comment by yosefe** [comment]
@MattBBaker can you post the ORNL failure?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2408/ for details.

**Comment by MattBBaker** [comment]
@yosefe @shamisp A lot of failures. And I'm seeing something similar on Titan.
```
[  FAILED  ] 88 tests, listed below:
[  FAILED  ] ud/test_ucp_fence32.atomic_add_fadd/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_fence32.atomic_add_fadd/1, where GetParam() = \ud
[  FAILED  ] udrc/test_ucp_fence32.atomic_add_fadd/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_fence32.atomic_add_fadd/0, where GetParam() = \ud,\rc
[  FAILED  ] cmrcx/test_ucp_fence32.atomic_add_fadd/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_fence32.atomic_add_fadd/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] shm/test_ucp_fence32.atomic_add_fadd/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_fence32.atomic_add_fadd/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] ud/test_ucp_fence64.atomic_add_fadd/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_fence64.atomic_add_fadd/1, where GetParam() = \ud
[  FAILED  ] udrc/test_ucp_fence64.atomic_add_fadd/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_fence64.atomic_add_fadd/0, where GetParam() = \ud,\rc
[  FAILED  ] cmrcx/test_ucp_fence64.atomic_add_fadd/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_fence64.atomic_add_fadd/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] shm/test_ucp_fence64.atomic_add_fadd/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_fence64.atomic_add_fadd/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] ud/test_ucp_atomic32.atomic_add/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_fadd/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_cswap/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_cswap/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_swap/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_add/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_swap/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic32.atomic_fadd/0, where GetParam() = \ud
[  FAILED  ] udrc/test_ucp_atomic32.atomic_add/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_fadd/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_swap/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_cswap/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_fadd/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_cswap/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_add/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic32.atomic_swap/1, where GetParam() = \ud,\rc
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_add/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_fadd/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_fadd/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_swap/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_cswap/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_add/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_cswap/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic32.atomic_swap/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] shm/test_ucp_atomic32.atomic_add/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_fadd/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_cswap/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_swap/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_add/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_fadd/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_swap/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic32.atomic_cswap/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] ud/test_ucp_atomic64.atomic_add/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_fadd/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_swap/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.unaligned_atomic_add/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_cswap/1, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_fadd/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_add/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.unaligned_atomic_add/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_swap/0, where GetParam() = \ud
[  FAILED  ] ud/test_ucp_atomic64.atomic_cswap/0, where GetParam() = \ud
[  FAILED  ] udrc/test_ucp_atomic64.atomic_cswap/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.unaligned_atomic_add/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_cswap/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.unaligned_atomic_add/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_fadd/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_fadd/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_add/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_swap/1, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_swap/0, where GetParam() = \ud,\rc
[  FAILED  ] udrc/test_ucp_atomic64.atomic_add/1, where GetParam() = \ud,\rc
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_add/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_swap/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_fadd/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_cswap/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_swap/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_cswap/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.unaligned_atomic_add/1, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.unaligned_atomic_add/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_fadd/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] cmrcx/test_ucp_atomic64.atomic_add/0, where GetParam() = \cm,\rc_mlx5
[  FAILED  ] shm/test_ucp_atomic64.atomic_add/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_swap/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_add/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_swap/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.unaligned_atomic_add/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_cswap/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_cswap/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.unaligned_atomic_add/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_fadd/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib
[  FAILED  ] shm/test_ucp_atomic64.atomic_fadd/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib
```
Example of a specific failure.
```
[----------] 10 tests from shm/test_ucp_atomic64
[ RUN      ] shm/test_ucp_atomic64.atomic_add/0
[1473881844.320257] [pod9:8057 :0]     ucp_worker.c:289  UCX  ERROR No transports capable of cpu atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_add/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib (51 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_swap/0
[1473881844.371882] [pod9:8057 :0]     ucp_worker.c:289  UCX  ERROR No transports capable of cpu atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_swap/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib (52 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_add/1
[1473881844.423358] [pod9:8057 :0]     ucp_worker.c:350  UCX  ERROR No transports capable of device atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_add/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib (51 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_swap/1
[1473881844.474845] [pod9:8057 :0]     ucp_worker.c:350  UCX  ERROR No transports capable of device atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_swap/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib (52 ms)
[ RUN      ] shm/test_ucp_atomic64.unaligned_atomic_add/0
[1473881844.526343] [pod9:8057 :0]     ucp_worker.c:289  UCX  ERROR No transports capable of cpu atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.unaligned_atomic_add/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib (51 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_cswap/1
[1473881844.577863] [pod9:8057 :0]     ucp_worker.c:350  UCX  ERROR No transports capable of device atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_cswap/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib (52 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_cswap/0
[1473881844.629391] [pod9:8057 :0]     ucp_worker.c:289  UCX  ERROR No transports capable of cpu atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_cswap/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib (51 ms)
[ RUN      ] shm/test_ucp_atomic64.unaligned_atomic_add/1
[1473881844.680925] [pod9:8057 :0]     ucp_worker.c:350  UCX  ERROR No transports capable of device atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.unaligned_atomic_add/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib (52 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_fadd/0
[1473881844.732627] [pod9:8057 :0]     ucp_worker.c:289  UCX  ERROR No transports capable of cpu atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_fadd/0, where GetParam() = \mm,\knem,\cma,\xpmem,ib (52 ms)
[ RUN      ] shm/test_ucp_atomic64.atomic_fadd/1
[1473881844.784318] [pod9:8057 :0]     ucp_worker.c:350  UCX  ERROR No transports capable of device atomic operations.
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/ucp/ucp_test.cc:238: Failure
Error: Unsupported operation
[  FAILED  ] shm/test_ucp_atomic64.atomic_fadd/1, where GetParam() = \mm,\knem,\cma,\xpmem,ib (51 ms)
[----------] 10 tests from shm/test_ucp_atomic64 (515 ms total)
```
```

**Comment by MattBBaker** [comment]
With recent push this works on Titan. 👍

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2409/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1745/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1746/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
failure is unrelated

**Pull request merged**